### PR TITLE
Added index for version vertex to improve performance

### PR DESCRIPTION
**Pull request by bsekar**
Signed-off-by: Sanjeev sanjeev.chopra@ge.com

### CODE REVIEW

**Comment by bsekar** [comment]
http://predix1.jenkins.build.ge.com/job/Predix-Security/job/acs-build-n-test-pullrequest/659/

**Comment by sanjeevchopra** [comment]
@amiri-ge  please review .
- PR checks passed - http://sjc1jenkins08.crd.ge.com:8080/job/Predix-Security/job/acs-build-n-test-pullrequest/663/console
- Siva verified performance tests on this branch as well

**Comment by sanjeevchopra** [comment]
@irinaepshteyn   , please review. (Dario is out)

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 1]
```diff
@@ -505,29 +506,14 @@ public void testGetResourceEntityAndDescendantsIds() {
...
-        assertThat(IteratorUtils.count(this.graph.vertices()), equalTo(1L));
-        assertThat(version, equalTo(2));
-        this.resourceRepository.setVersion(3);
-        this.resourceRepository.setVersion(4);
-        assertThat(IteratorUtils.count(this.graph.vertices()), equalTo(1L));
-        version = this.resourceRepository.getVersion();
-        assertThat(version, equalTo(4));
+        
+        this.resourceRepository.createVersionVertex(2);
+        assertThat(IteratorUtils.count(this.graph.vertices()), equalTo(2L));
```
Comment (bhuvan-rk) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java:507]: @sanjeevchopra 
Think we need to add assertion that checkVersionVertexExists(2) returns false before creation and returns true after being created.

**Review by bhuvan-rk — CHANGES_REQUESTED** [review]
@sanjeevchopra 
Done reviewing.

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java (hunk 1 above)
Comment (sanjeevchopra) [reply service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java:507]: @bhuvan-rk this is testing that a new vertex is added. the exists test is being done with version 1.  Though, a exists == false check should be added before version 1 is created.

File: service/src/main/java/com/ge/predix/acs/config/GraphConfig.java [hunk 2]
```diff
@@ -93,17 +96,27 @@ public Graph createGraph() throws InterruptedException {
...
-                RESOURCE_ID_KEY);
-        createTwoKeyUniqueCompositeIndex(newGraph, BY_ZONE_AND_SUBJECT_UNIQUE_INDEX_NAME, ZONE_ID_KEY, SUBJECT_ID_KEY);
-        createEdgeIndex(newGraph, BY_SCOPE_INDEX_NAME, PARENT_EDGE_LABEL, SCOPE_PROPERTY_KEY);
+        createUniqueCompositeIndex(newGraph, BY_ZONE_AND_RESOURCE_UNIQUE_INDEX_NAME, new String[] {ZONE_ID_KEY,
+                RESOURCE_ID_KEY});
 
-        LOGGER.info("Initialized titan db.");
-        return newGraph;
+        createUniqueCompositeIndex(newGraph, BY_ZONE_AND_SUBJECT_UNIQUE_INDEX_NAME,
+                new String[] { ZONE_ID_KEY, SUBJECT_ID_KEY });
```
Comment (DarioAmiri) [inline service/src/main/java/com/ge/predix/acs/config/GraphConfig.java:114]: nitpick: why is the formatting of this call to createUniqueCompositeIndex different than the one before? Are we using the eclipse formatting rules for this project?

File: service/src/main/java/com/ge/predix/acs/config/GraphConfig.java [hunk 3]
```diff
@@ -135,38 +148,46 @@ public static void createUniqueIndexForLabel(final Graph newGraph, final String
...
 
-    public static void createTwoKeyUniqueCompositeIndex(final Graph newGraph, final String indexName,
-            final String indexKey1, final String indexKey2) throws InterruptedException {
+    public static void createUniqueCompositeIndex(final Graph newGraph, final String indexName,
+            final String[] propertyKeyNames) throws InterruptedException {
+        createUniqueCompositeIndex(newGraph, indexName, propertyKeyNames, String.class);
+    }
+
+    public static void createUniqueCompositeIndex(final Graph newGraph, final String indexName,
+            final String[] propertyKeyNames, final Class<?> propertyKeyType) throws InterruptedException {
```
Comment (DarioAmiri) [inline service/src/main/java/com/ge/predix/acs/config/GraphConfig.java:157]: This method signature assumes that all property keys of the composite index are the same type. Will this always be the case? Does it make sense to provide property (keyName, keyType) tuples instead?

**Review by DarioAmiri — APPROVED** [review]
I found minor issues to consider addressing in a future story.

File: service/src/main/java/com/ge/predix/acs/config/GraphConfig.java (hunk 3 above)
Comment (sanjeevchopra) [reply service/src/main/java/com/ge/predix/acs/config/GraphConfig.java:157]: yes, we can add that when we have that use case.

File: service/src/main/java/com/ge/predix/acs/config/GraphConfig.java (hunk 2 above)
Comment (sanjeevchopra) [reply service/src/main/java/com/ge/predix/acs/config/GraphConfig.java:114]: fixed

**Review by sanjeevchopra — COMMENTED** [review]
@amiri-ge @bhuvan-rk 
Addressed your comments and removed  redundant graph setup in tests.

**Pull request merged**